java-security/src/main/java/com/sap/cloud/security/token/validation/validators/JwtCnfValidator.java

```diff
@@ -0,0 +1,83 @@
+package com.sap.cloud.security.token.validation.validators;
+
+import com.sap.cloud.security.json.JsonObject;
+import com.sap.cloud.security.token.SecurityContext;
+import com.sap.cloud.security.token.Token;
+import com.sap.cloud.security.token.TokenClaims;
+import com.sap.cloud.security.token.validation.ValidationResult;
+import com.sap.cloud.security.token.validation.ValidationResults;
+import com.sap.cloud.security.token.validation.Validator;
+import com.sap.cloud.security.xsuaa.mtls.X509Parser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+
+/**
+ * Validates if the jwt access token is intended for the OAuth2 client of this
+ * application, in certificate based authentication scenario. The cnf claim
+ * provides the SHA-256 thumbprint of the X.509 certificate information, this
+ * information needs to match the thumbprint of the provided certificate.
+ * <p>
+ * Validates whether there 'cnf' thumbprint value matches with the X509
+ * certificate from the request.
+ */
+public class JwtCnfValidator implements Validator<Token> {
+	private static final Logger LOGGER = LoggerFactory.getLogger(JwtCnfValidator.class);
+	private final String trustedClientId;
+
+	public JwtCnfValidator(String clientId) {
+		trustedClientId = clientId;
+	}
+
+	/**
+	 * Validates the cnf thumbprint of X509 certificate against trusted
+	 * certificate's thumbprint.
+	 * 
+	 * @param token
+	 *            token to be validated
+	 * @return validation result. Result is valid when both thumbprints match.
+	 */
+	@Override
+	public ValidationResult validate(Token token) {
+
+		String cnf = extractCnfThumbprintFromToken(token);
+		LOGGER.info("Cnf thumbprint: {}", cnf);
+		if (cnf == null && token.getAudiences().size() == 1 && token.getAudiences().contains(trustedClientId)) {
```

Comment:
for this check pls move cnf check into the else block... and pls check whether the 3rd check aud == trustedClientId is required (will this validator be executed in case the audience validation (which comes first) failed?

Reply:
if aud is empty check azp with getAudiences method

Reply:
@nenaraab  I've checked that and only in case of Xsuaa azp is added to audience list and then validated -> [JwtAudienceValidator](https://github.com/SAP/cloud-security-xsuaa-integration/blob/d1fe8fde82997f505ae6aea4252333c26c82591b/java-security/src/main/java/com/sap/cloud/security/token/validation/validators/JwtAudienceValidator.java#L100)

 So are you sure that we can accept azp also in case of IAS? in that case I would adapt the audience validator.